src/wmtk/TriMesh.h

```diff
@@ -300,38 +300,113 @@ class TriMesh
     size_t get_next_empty_slot_t()
     {
         m_tri_connectivity.emplace_back();
+        resize_attributes(
+            m_vertex_connectivity.size(),
+            m_tri_connectivity.size() * 3,
+            m_tri_connectivity.size());
         return m_tri_connectivity.size() - 1;
     }
     size_t get_next_empty_slot_v()
     {
         m_vertex_connectivity.emplace_back();
+        resize_attributes(
+            m_vertex_connectivity.size(),
+            m_tri_connectivity.size() * 3,
+            m_tri_connectivity.size());
         return m_vertex_connectivity.size() - 1;
     }
 
 protected:
     virtual bool split_before(const Tuple& t) { return true; }
     virtual bool split_after(const Tuple& t) { return true; }
-    virtual bool collapse_before(const Tuple& t) { return true; }
-    virtual bool collapse_after(const Tuple& t) { return true; }
+    // check link, check if it's the last edge
+    virtual bool collapse_before(const Tuple& t)
+    {
+        if (check_link_condition(t) && check_manifold(t)) return true;
```

Comment:
remove `check_manifold` change `check_link_condition` to include open boundaries
test `check_link_condition` vs libigl